### PR TITLE
Add support to TOP ECS on DNS Handler

### PR DESCRIPTION
**Pull request by leoparente**

### CODE REVIEW

File: src/handlers/dns/DnsAdditionalRecord.h [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+    ecs->source_netmask = iterator[6];
+    ecs->scope_netmask = iterator[7];
+
+    if (ecs->family == FamilyAddressEnum::IPV4) {
+        char addrBuffer[INET_ADDRSTRLEN];
+        if (inet_ntop(AF_INET, &iterator[8], addrBuffer, sizeof(addrBuffer)) != NULL)
+            ecs->client_subnet = addrBuffer;
+    } else if (ecs->family == FamilyAddressEnum::IPV6) {
+        char addrBuffer[INET6_ADDRSTRLEN];
+        if (inet_ntop(AF_INET6, &iterator[8], addrBuffer, sizeof(addrBuffer)) != NULL)
```
Comment (leoparente) [inline src/handlers/dns/DnsAdditionalRecord.h]: @rboucher-ns1 @weyrick I need a help here. `inet_ntop` is returing "2001:470:1f0b:1600:2000::" but it should be "2001:470:1f0b:1600::" based on what wireshark and also the byte array: 20 01 04 70 1f 0b 16;

Comment (leoparente) [reply src/handlers/dns/DnsAdditionalRecord.h]: If you need to check the packet that I'm using you can download the ecs.pcap that I added

Comment (leoparente) [reply src/handlers/dns/DnsAdditionalRecord.h]: I was able to fix it by ensuring that I have a ipv6 size vector (16 bytes) with zero initialized values

**Comment by leoparente** [comment]
@weyrick should we add TOP ECS to a group not enabled by default?

File: src/handlers/dns/DnsAdditionalRecord.h (hunk 1 above)
Comment (leoparente) [reply src/handlers/dns/DnsAdditionalRecord.h]: It seems it was trying to get full size, and was getting garbage from memory.

**Comment by weyrick** [comment]
> @weyrick should we add TOP ECS to a group not enabled by default?

yes, not enabled by default

File: src/handlers/dns/DnsStreamHandler.h [hunk 2]
```diff
@@ -112,6 +113,7 @@ class DnsMetricsBucket final : public visor::AbstractMetricsBucket
         : _dnsXactFromTimeUs("dns", {"xact", "out", "quantiles_us"}, "Quantiles of transaction timing (query/reply pairs) when host is client, in microseconds")
         , _dnsXactToTimeUs("dns", {"xact", "in", "quantiles_us"}, "Quantiles of transaction timing (query/reply pairs) when host is server, in microseconds")
         , _dns_qnameCard("dns", {"cardinality", "qname"}, "Cardinality of unique QNAMES, both ingress and egress")
+        , _dns_topECS("dns", "ecs", {"top_ecs"}, "Top EDNS Client Subnet (ECS) on the query side of a transaction")
```
Comment (weyrick) [inline src/handlers/dns/DnsStreamHandler.h]: as discussed, let's go with "top_ecs_query"

File: src/handlers/dns/DnsLayer.cpp [hunk 3]
```diff
@@ -190,8 +190,11 @@ bool DnsLayer::parseResources(bool queryOnly)
             m_FirstAnswer = newResource;
         else if (resType == DnsAuthorityType && m_FirstAuthority == NULL)
             m_FirstAuthority = newResource;
-        else if (resType == DnsAdditionalType && m_FirstAdditional == NULL)
+        else if (resType == DnsAdditionalType && m_FirstAdditional == NULL) {
             m_FirstAdditional = newResource;
+            if (additionalOnly)
```
Comment (weyrick) [inline src/handlers/dns/DnsLayer.cpp]: please use braces

File: src/handlers/dns/DnsAdditionalRecord.h [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+    ecs->scope_netmask = array[7];
+    size_t offset = 8;
+
+    if (ecs->family == FamilyAddressEnum::IPV4) {
+        char addrBuffer[INET_ADDRSTRLEN];
+        std::array<uint8_t, IPV4_BYTE_SIZE> ipv4 = {};
+        for (auto i = offset; i < std::min(size, IPV4_BYTE_SIZE); i++) {
+            ipv4[i - offset] = array[i];
+        }
+        if (inet_ntop(AF_INET, &ipv4, addrBuffer, sizeof(addrBuffer)) != NULL)
```
Comment (weyrick) [inline src/handlers/dns/DnsAdditionalRecord.h]: please use braces

File: src/handlers/dns/DnsAdditionalRecord.h [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+            ecs->client_subnet = addrBuffer;
+    } else if (ecs->family == FamilyAddressEnum::IPV6) {
+        char addrBuffer[INET6_ADDRSTRLEN];
+        std::array<uint8_t, IPV6_BYTE_SIZE> ipv6 = {};
+        for (auto i = offset; i < std::min(size, IPV6_BYTE_SIZE); i++) {
+            ipv6[i - offset] = array[i];
+        }
+        if (inet_ntop(AF_INET6, &ipv6, addrBuffer, sizeof(addrBuffer)) != NULL)
```
Comment (weyrick) [inline src/handlers/dns/DnsAdditionalRecord.h]: braces

**Review by weyrick — CHANGES_REQUESTED** [review]

File: src/handlers/dns/DnsStreamHandler.h (hunk 2 above)
Comment (leoparente) [reply src/handlers/dns/DnsStreamHandler.h]: done

File: src/handlers/dns/DnsLayer.cpp (hunk 3 above)
Comment (leoparente) [reply src/handlers/dns/DnsLayer.cpp]: done

File: src/handlers/dns/DnsAdditionalRecord.h (hunk 4 above)
Comment (leoparente) [reply src/handlers/dns/DnsAdditionalRecord.h]: done

File: src/handlers/dns/DnsAdditionalRecord.h (hunk 5 above)
Comment (leoparente) [reply src/handlers/dns/DnsAdditionalRecord.h]: done

**Comment by leoparente** [comment]
I've started doing some profiling using real scenario pcap data and database.

The criteria to profile were:
**1)** use pktvisor-reader with same configuration in all profiling tests
**2)** Start process and wait at least 1M packets to be processed 

The result for `develop branch` was the following:
geo_develop
We can notice that _Net Handler Call_ used **43.6%** while DNS Handler call consumed 28.3% of process CPU.

If we take a look on Net Handler Calls, MaxmindDB calls consumes **28.7%** (City + ASN)
geo_dev_expand

File: src/handlers/dns/DnsStreamHandler.h [hunk 6]
```diff
@@ -112,6 +114,7 @@ class DnsMetricsBucket final : public visor::AbstractMetricsBucket
         : _dnsXactFromTimeUs("dns", {"xact", "out", "quantiles_us"}, "Quantiles of transaction timing (query/reply pairs) when host is client, in microseconds")
         , _dnsXactToTimeUs("dns", {"xact", "in", "quantiles_us"}, "Quantiles of transaction timing (query/reply pairs) when host is server, in microseconds")
         , _dns_qnameCard("dns", {"cardinality", "qname"}, "Cardinality of unique QNAMES, both ingress and egress")
+        , _dns_topECSQuery("dns", "ecs", {"top_ecs_query"}, "Top EDNS Client Subnet (ECS) on the query side of a transaction")
```
Comment (rboucher-me) [inline src/handlers/dns/DnsStreamHandler.h]: `top_query_ecs`
"Top EDNS Client Subnets (ECS) observed in DNS queries"

File: src/handlers/dns/DnsStreamHandler.h [hunk 7]
```diff
@@ -158,6 +161,7 @@ class DnsMetricsBucket final : public visor::AbstractMetricsBucket
     void to_prometheus(std::stringstream &out, Metric::LabelMap add_labels = {}) const override;
     void update_topn_metrics(size_t topn_count) override
     {
+        _dns_topECSQuery.set_topn_count(topn_count);
```
Comment (rboucher-me) [inline src/handlers/dns/DnsStreamHandler.h]: Should we change `_dns_topECSQuery` to `_dns_topQueryECS` to line up with the metric name?

File: src/handlers/dns/tests/window-schema.json [hunk 8]
```diff
@@ -15,6 +15,7 @@
         "length": 31,
         "start_ts": 1614874231
       },
+      "top_ecs_query": [],
```
Comment (rboucher-me) [inline src/handlers/dns/tests/window-schema.json]: `top_query_ecs`

File: src/handlers/dns/tests/window-schema.json [hunk 9]
```diff
@@ -368,6 +369,7 @@
       "required": [
         "cardinality",
         "period",
+        "top_ecs_query",
```
Comment (rboucher-me) [inline src/handlers/dns/tests/window-schema.json]: `top_query_ecs`

File: src/handlers/dns/tests/window-schema.json [hunk 10]
```diff
@@ -448,6 +450,20 @@
           },
           "additionalProperties": false
         },
+        "top_ecs_query": {
```
Comment (rboucher-me) [inline src/handlers/dns/tests/window-schema.json]: `top_query_ecs`

File: src/handlers/dns/tests/window-schema.json [hunk 11]
```diff
@@ -448,6 +450,20 @@
...
+          "$id": "#/properties/dns/properties/top_ecs_query",
```
Comment (rboucher-me) [inline src/handlers/dns/tests/window-schema.json]: `top_query_ecs`

File: src/handlers/dns/tests/window-schema.json [hunk 12]
```diff
@@ -448,6 +450,20 @@
...
+          "type": "array",
+          "title": "The top_ecs_query schema",
```
Comment (rboucher-me) [inline src/handlers/dns/tests/window-schema.json]: `top_query_ecs`

File: src/handlers/dns/tests/window-schema.json [hunk 13]
```diff
@@ -448,6 +450,20 @@
...
+          "description": "An explanation about the purpose of this instance.",
+          "default": [],
+          "examples": [
+            []
+          ],
+          "additionalItems": true,
+          "items": {
+            "$id": "#/properties/dns/properties/top_ecs_query/items"
```
Comment (rboucher-me) [inline src/handlers/dns/tests/window-schema.json]: `top_query_ecs`

**Review by rboucher-me — CHANGES_REQUESTED** [review]

File: src/handlers/dns/DnsStreamHandler.h (hunk 6 above)
Comment (leoparente) [reply src/handlers/dns/DnsStreamHandler.h]: done

File: src/handlers/dns/DnsStreamHandler.h (hunk 7 above)
Comment (leoparente) [reply src/handlers/dns/DnsStreamHandler.h]: agreed. done

File: src/handlers/dns/tests/window-schema.json (hunk 9 above)
Comment (leoparente) [reply src/handlers/dns/tests/window-schema.json]: done

**Review by rboucher-me — APPROVED** [review]
Looks good

**Review by weyrick — APPROVED** [review]

**Pull request merged**